presto-raptor/src/main/java/com/facebook/presto/raptor/RaptorPageSink.java

```diff
@@ -144,21 +154,14 @@ private Page createPageWithSampleWeightBlock(Page page, Block sampleWeightBlock)
      * <li>pageBuffer has more than Integer.MAX_VALUE rows (PagesSorter.sort can sort Integer.MAX_VALUE rows at a time)</li>
      * </ul>
      */
-    private void flushPageBufferIfNecessary(int rowsToAdd)
+    private boolean shouldFlushPageBuffer(int rowsToAdd)
     {
-        if (storagePageSink.isFull()) {
-            // This StoragePageSink is full, flush it for the next batch of pages
-            flushPages(pageBuffer.getPages());
-            pageBuffer.reset();
-            storagePageSink.flush();
-            return;
-        }
+        return storagePageSink.isFull() || pageBuffer.isFull() || (!sortFields.isEmpty() && (rowsToAdd > allowedRowsToAdd()));
```

Comment:
I think it's safe to say that a page buffer can hold at most 2^31 pages, even if not sorting. Just keeps things simpler.